internal/graph/check_fast_path.go

```diff
@@ -0,0 +1,668 @@
+		}
+		close(iterChan)
+		cleanupIteratorProducers(iterProducers)
+	}()
+	/*
+		collect iterators from all channels, once none are nil
+		start performing intersection algorithm across the heads, if an iterator is drained
+		ask to see if the channel has a new iterator, otherwise consider it done
+		exit if one of the channels closes as there is no more possible intersection of all
+	*/
+
+	childrenTotal := len(iterProducers)
+	for len(iterProducers) == childrenTotal {
+		if ctx.Err() != nil {
+			return
+		}
+		var err error
+		iterProducers, err = pollIteratorProducers(ctx, iterProducers)
+		if err != nil {
+			concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+			return
+		}
+		if len(iterProducers) != childrenTotal {
+			// short circuit
+			return
+		}
+
+		maxKey := ""
+		indexes := make([]int, 0)
+		allIters := true
+		for idx, producer := range iterProducers {
+			v, err := producer.iter.Head(ctx)
+			if err != nil {
+				if storage.IterIsDoneOrCancelled(err) {
+					producer.iter.Stop()
+					producer.iter = nil
+					allIters = false
+					// we need to ensure we have all iterators at all times
+					break
+				}
+				concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+				return
+			}
+
+			if idx == 0 {
+				maxKey = v.GetObject()
+			}
+
+			if maxKey == v.GetObject() {
+				indexes = append(indexes, idx)
+			}
+
+			if maxKey < v.GetObject() {
+				maxKey = v.GetObject()
+				indexes = []int{idx}
+			}
+		}
+		if !allIters {
+			// we need to ensure we have all iterators at all times
+			continue
+		}
+
+		// all children have the same value
+		if len(indexes) == childrenTotal {
+			// all entries are the same thus flush entry and move iterators
+			// there should only be 1 value
+			for idx, iterIdx := range indexes {
+				t, err := iterProducers[iterIdx].iter.Next(ctx)
+				if err != nil {
+					if storage.IterIsDoneOrCancelled(err) {
+						iterProducers[iterIdx].iter.Stop()
+						iterProducers[iterIdx].iter = nil
+						break
+					}
+					concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+					return
+				}
+				if idx == 0 {
+					batch = append(batch, t)
+				}
+			}
+
+			if len(batch) > IteratorBatchThreshold {
+				concurrency.TrySendThroughChannel(ctx, &iteratorMsg{iter: storage.NewStaticTupleKeyIterator(batch)}, iterChan)
+				batch = make([]*openfgav1.TupleKey, 0)
+			}
+			continue
+		}
+
+		// move all entries to less than the MAX to be >= than MAX
+		for _, producer := range iterProducers {
+			t, err := producer.iter.Head(ctx)
+			if err != nil {
+				if storage.IterIsDoneOrCancelled(err) {
+					producer.iter.Stop()
+					producer.iter = nil
+					continue
+				}
+				concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+				return
+			}
+			tmpKey := t.GetObject()
+			for tmpKey < maxKey {
+				t, err := producer.iter.Next(ctx)
+				if err != nil {
+					if storage.IterIsDoneOrCancelled(err) {
+						producer.iter.Stop()
```

Comment:
Do we have a bug here?  Assume you have the following
A=[5]
B=[6 7]
C=[6 7]

max key = 6.  When we process A, we would have call next and the iterator will be done.  We should have return at this point.  Otherwise, if we continue, the pollIteratorProducers would have trimmed slice A and we think there is intersection for 7.

Reply:
Discussed offline - there is no bug because we will exit due to short circuit on len(iterProducers) != childrenTotal